**REALGenomeSIM** simulates GWAS data with **R**ealistic **E**pistatic effects, **A**dditive effects, and **L**D patterns with **Genome** **S**ubsampling and **I**ntegrated **M**odeling. It divides each simulated genome into chunks such that the probability that any genomic position demarates two chunks equals the probability that this position would demarcate a real (given) recombination event. Each chunk is then filled with the homologous genomic subset of a real whole genome that is sampled uniformly (with replacement) from an input plink (.bed, .bim, .fam) fileset. Triadsim has applied this method Simulate new trios from a dataset of real trios, and the resulting simulated datasets produce flawless LD curves. REALGenomeSIM also simulates equally realistic data, but it also adds several new features and addresses several inadequacies from the original Triadsim Algorithm. 

<p align="center">
<img src="https://github.com/EpistasisLab/REALGenomeSIM/blob/master/images/REALGenomeSIM.png" width=1500/>
</p>
